core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ast/planNodes/Select.java

```diff
@@ -169,13 +171,28 @@ public boolean equals(Object o) {
 			return false;
 		}
 		Select select = (Select) o;
-		return connection.equals(select.connection) &&
-				mapper.equals(select.mapper) &&
-				query.equals(select.query);
+		if (connection instanceof MemoryStoreConnection && select.connection instanceof MemoryStoreConnection) {
```

Comment:
Why are we special-casing the MemoryStoreConnection here?

Reply:
Because every plan node that needs a connection to the two internal transaction memory stores (one for added one for removed) ask for a new connection every time instead of sharing one. This means that we need to check the base sail instead of the actual connection when comparing two plan nodes to see if they are equal.

Reply:
And by equal I mean "will produce the same results".